src/include/storage/data_table.h

```diff
@@ -74,7 +95,12 @@ class DataTable {
      * @return if the two iterators point to the same slot
      */
     bool operator==(const SlotIterator &other) const {
-      // TODO(Tianyu): I believe this is enough?
+      if (LIKELY(other.is_end_)) {
+        return block_index_ >= end_index_;
+      }
+      if (LIKELY(is_end_)) {
+        return other.block_index_ >= other.end_index_;
+      }
       return current_slot_ == other.current_slot_;
```

Comment:
So I should definitely add a comment but this is done to avoid computation when comparing to end(). Since the data table only needs to provide snap shot isolation, when the `SlotIterator` is created, it only needs to iterate through all the slots that are available then. So the end index is kept and maintained within each iterator that tells it when its end is. This essentially means that we don't need to do anything in end to see when the iterator has reached the end. This is why we have the magic `end_` constant. Whenever an iterator is being compared to `end_` it knows to just check if it has reached the end of the `DataTable` when it was created.

Reply:
So the magic `end_` constant is a static value common to all `DataTable`s and is never updated.  I don't have an issue with that concept (it matches STL iterators in practice).   My point is that we should have a single canonical value for the tupleslot of an invalid/end iterator (matching the STL) which makes equality comparisons cheap (a single, unconditional comparison).

Reply:
Im not sure I understand the difference?

Reply:
Thanks for making the change.  If you still want to discuss why I asked for this, hit me up on Slack and we figure out a time to chat (I think it would be easier to hash out in real time).

Reply:
No it makes a lot of sense. I think it might also make it faster since it makes equal comparisons much faster probably.